File: 回溯算法/04组合总和3.py
# 216. 组合总和 III

def combinationSum3(k: int, n: int):
    
    res = []
    track = []



    def backtrack(track, k, n, start):
        sum_track = sum(track)
        if len(track) == k and sum_track == n:  
            res.append(track.copy())
        elif sum_track > n:  # 还未加到K个数，总和已超过n，此时无需再往后尝试。因此，此处做剪枝
            return

        for i in range(start, 10):
            if i in track:
                continue

            track.append(i)

            backtrack(track, k, n, i+1)  # i+1 避免重复添加

            track.pop(-1)

    backtrack([], k, n, 1)
    return res

File: 回溯算法/04组合总和3.java
class Solution {
    List<List<Integer>> res = new ArrayList<>();
    List<Integer> temp = new ArrayList<>();

    /**
     * 216. 组合总和3
     *
     * @param k
     * @param n
     * @return
     */
    public List<List<Integer>> combinationSum3(int k, int n) {
        drawBack(n, k, 0, 1);
        return res;
    }

    private void drawBack(int n, int k, int val, int start) {
        if (temp.size() == k && val == n) {
            res.add(new ArrayList<>(temp));
            return;
        }
        // 组合超过K，不符合题意
        if (temp.size() > k)
            return;
        // 重点：此处的剪枝操作
        // i <= n-(k-temp.size())+1
        for (int i = start; i <= 9; i++) {
            // 选择
            temp.add(i);
            val += i;

            drawBack(n, k, val, i+1);

            // 回撤选择
            val -= i;
            temp.remove(temp.size()-1);
        }
    }
}